Trying to set up local development, but can't run YAML tests

Hi! I heard about this library from typing-sig and it looks really interesting. I tried setting up local development and ran into a couple issues.

1. When I run `pre-commit run --all`, there's a flake8 error:

```
(env) ➜  phantom-types git:(main) ✗ pre-commit run --all
Check for case conflicts.................................................Passed
Check for merge conflicts................................................Passed
Fix End of Files.........................................................Passed
Trim Trailing Whitespace.................................................Passed
Debug Statements (Python)................................................Passed
Detect Private Key.......................................................Passed
pyupgrade................................................................Passed
autoflake................................................................Passed
isort....................................................................Passed
black....................................................................Passed
blacken-docs.............................................................Passed
flake8...................................................................Failed
- hook id: flake8
- exit code: 1

src/phantom/fn.py:36:9: B018 Found useless expression. Either assign it to a variable or remove it.

Validate GitHub Workflows................................................Passed
mypy.....................................................................Passed
check-manifest...........................................................Passed
format-readme............................................................Failed
- hook id: format-readme
- exit code: 1

Executable `docker` not found

(env) ➜  phantom-types git:(main) ✗
```

And indeed, looking at [line 36 of that file](https://github.com/antonagestam/phantom-types/blob/51b421899321b2a381f2bf02a74d30cc61c2467c/src/phantom/fn.py#L36) shows an unassigned string expression. I assume this is just leftover since that line hasn't been touched since April, but maybe I'm misunderstanding.

2. Running `make test` and/or `pytest` fails on all of the yaml-based tests. Here's an example:

```
(env) ➜  phantom-types git:(main) ✗ pytest tests/ext/test_phonenumbers.yaml
============================================================================== test session starts ==============================================================================
platform darwin -- Python 3.9.9, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /Users/tushar/code/phantom-types, configfile: pyproject.toml
plugins: mypy-plugins-1.9.2, typeguard-2.13.3
collected 3 items

tests/ext/test_phonenumbers.yaml FFF                                                                                                                                      [100%]

=================================================================================== FAILURES ====================================================================================
_____________________________________________________________________________ bound_is_not_subtype ______________________________________________________________________________
/Users/tushar/code/phantom-types/tests/ext/test_phonenumbers.yaml:9:
E   pytest_mypy_plugins.utils.TypecheckAssertionError: Invalid output:
E   Actual:
E     ../../../../../../Users/tushar/code/phantom-types/src/phantom/base:53: error: Unused "type: ignore[misc]" comment (diff)
E     ../../../../../../Users/tushar/code/phantom-types/src/phantom/base:144: error: ClassVar cannot contain type variables  [misc] (diff)
E     ../../../../../../Users/tushar/code/phantom-types/src/phantom/base:162: error: Unused "type: ignore" comment (diff)
E     main:7: error: Argument 1 to "takes_phone_number" has incompatible type "str"; expected "PhoneNumber"  [arg-type] (diff)
E     main:12: error: Argument 1 to "takes_formatted_phone_number" has incompatible type "str"; expected "FormattedPhoneNumber"  [arg-type] (diff)
E   Expected:
E     main:7: error: Argument 1 to "takes_phone_number" has incompatible type "str"; expected "PhoneNumber"  [arg-type] (diff)
E     main:12: error: Argument 1 to "takes_formatted_phone_number" has incompatible type "str"; expected "FormattedPhoneNumber"  [arg-type] (diff)
E   Alignment of first line difference:
E     E: main:7: error: Argument 1 to "takes_phone_number" has incompatible type ...
E     A: ../../../../../../Users/tushar/code/phantom-types/src/phantom/base:53: e...
E        ^
________________________________________________________________________________ can_instantiate ________________________________________________________________________________
/Users/tushar/code/phantom-types/tests/ext/test_phonenumbers.yaml:16:
E   pytest_mypy_plugins.utils.TypecheckAssertionError: Output is not expected:
E   Actual:
E     ../../../../../../Users/tushar/code/phantom-types/src/phantom/base:53: error: Unused "type: ignore[misc]" comment (diff)
E     ../../../../../../Users/tushar/code/phantom-types/src/phantom/base:144: error: ClassVar cannot contain type variables  [misc] (diff)
E     ../../../../../../Users/tushar/code/phantom-types/src/phantom/base:162: error: Unused "type: ignore" comment (diff)
E   Expected:
E     (empty)
___________________________________________________________________________________ can_infer ___________________________________________________________________________________
/Users/tushar/code/phantom-types/tests/ext/test_phonenumbers.yaml:30:
E   pytest_mypy_plugins.utils.TypecheckAssertionError: Output is not expected:
E   Actual:
E     ../../../../../../Users/tushar/code/phantom-types/src/phantom/base:53: error: Unused "type: ignore[misc]" comment (diff)
E     ../../../../../../Users/tushar/code/phantom-types/src/phantom/base:144: error: ClassVar cannot contain type variables  [misc] (diff)
E     ../../../../../../Users/tushar/code/phantom-types/src/phantom/base:162: error: Unused "type: ignore" comment (diff)
E   Expected:
E     (empty)
============================================================================ short test summary info ============================================================================
FAILED tests/ext/test_phonenumbers.yaml::bound_is_not_subtype -
FAILED tests/ext/test_phonenumbers.yaml::can_instantiate -
FAILED tests/ext/test_phonenumbers.yaml::can_infer -
=============================================================================== 3 failed in 0.81s ===============================================================================
(env) ➜  phantom-types git:(main) ✗
```

It looks to me like every test is throwing errors on lines 53, 144, and 162 of `src/phantom/base.py`, e.g., [here](https://github.com/antonagestam/phantom-types/blob/main/src/phantom/base.py#L53). If I run `mypy .`, the same errors appear in the output there.

The first test example (`bound_is_not_subtype`) shows that I _am_ getting the expected results, but every test has these three extra errors appended to it. I assume this is a difference in mypy config that I'm missing, but I'd love help figuring this out.

Thanks for your great work on this library!

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -11,7 +11,7 @@ repos:
       - id: debug-statements
       - id: detect-private-key
   - repo: https://github.com/asottile/pyupgrade
-    rev: v2.30.1
+    rev: v2.31.0
     hooks:
       - id: pyupgrade
         args:
@@ -46,11 +46,11 @@ repos:
           - flake8-comprehensions
           - flake8-tidy-imports
   - repo: https://github.com/sirosen/check-jsonschema
-    rev: 0.9.1
+    rev: 0.10.0
     hooks:
       - id: check-github-workflows
   - repo: https://github.com/pre-commit/mirrors-mypy
-    rev: v0.930
+    rev: v0.931
     hooks:
       - id: mypy
         args: []

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -58,8 +58,9 @@ repos:
         additional_dependencies:
           - pytest
           - typeguard
-          - phonenumbers>=8.12.33
+          - phonenumbers>=8.12.41
           - pydantic
+          - types-setuptools
   - repo: https://github.com/mgedmin/check-manifest
     rev: "0.47"
     hooks:

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -2,7 +2,7 @@ default_language_version:
   python: python3.10
 repos:
   - repo: https://github.com/pre-commit/pre-commit-hooks
-    rev: v4.0.1
+    rev: v4.1.0
     hooks:
       - id: check-case-conflict
       - id: check-merge-conflict
@@ -11,7 +11,7 @@ repos:
       - id: debug-statements
       - id: detect-private-key
   - repo: https://github.com/asottile/pyupgrade
-    rev: v2.27.0
+    rev: v2.30.1
     hooks:
       - id: pyupgrade
         args:
@@ -25,15 +25,15 @@ repos:
           - --remove-all-unused-imports
           - --ignore-init-module-imports
   - repo: https://github.com/pycqa/isort
-    rev: 5.9.3
+    rev: 5.10.1
     hooks:
       - id: isort
   - repo: https://github.com/psf/black
-    rev: 21.9b0
+    rev: 21.12b0
     hooks:
       - id: black
   - repo: https://github.com/asottile/blacken-docs
-    rev: v1.11.0
+    rev: v1.12.0
     hooks:
       - id: blacken-docs
         additional_dependencies: [black==21.9b0]
@@ -46,11 +46,11 @@ repos:
           - flake8-comprehensions
           - flake8-tidy-imports
   - repo: https://github.com/sirosen/check-jsonschema
-    rev: 0.4.1
+    rev: 0.9.1
     hooks:
       - id: check-github-workflows
   - repo: https://github.com/pre-commit/mirrors-mypy
-    rev: v0.910
+    rev: v0.930
     hooks:
       - id: mypy
         args: []

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -8,4 +8,4 @@ target-version = ["py38"]
 [tool.pytest.ini_options]
 norecursedirs = ["examples"]
 testpaths = ["tests", "src", "docs"]
-addopts = "--mypy-ini-file=setup.cfg --doctest-modules"
+addopts = "--mypy-ini-file=setup.cfg --mypy-only-local-stub --doctest-modules"

--- a/setup.cfg
+++ b/setup.cfg
@@ -47,7 +47,6 @@ test =
     pytest
     pytest-mypy-plugins>=1.9.3
     coverage
-    types-setuptools
 
 [flake8]
 exclude = appveyor,.idea,.git,.venv,.tox,__pycache__,*.egg-info,build

--- a/setup.cfg
+++ b/setup.cfg
@@ -47,6 +47,7 @@ test =
     pytest
     pytest-mypy-plugins>=1.9.3
     coverage
+    types-setuptools
 
 [flake8]
 exclude = appveyor,.idea,.git,.venv,.tox,__pycache__,*.egg-info,build
@@ -91,9 +92,6 @@ disallow_untyped_calls = False
 disallow_any_expr = False
 disallow_untyped_decorators = False
 disallow_incomplete_defs = False
-
-[mypy-typeguard.*]
-ignore_errors = True
 
 [coverage:run]
 source = src

--- a/setup.cfg
+++ b/setup.cfg
@@ -45,9 +45,7 @@ pydantic =
 test =
     mypy>=0.930
     pytest
-    # todo: Restore to mainline pytest-mypy-plugins once PR is closed.
-    #       https://github.com/typeddjango/pytest-mypy-plugins/pull/92
-    pytest-mypy-plugins @ https://github.com/antonagestam/pytest-mypy-plugins/archive/01957984e26118c9b176e3d055c858e1db029cf1.tar.gz
+    pytest-mypy-plugins>=1.9.3
     coverage
 
 [flake8]

--- a/setup.cfg
+++ b/setup.cfg
@@ -43,9 +43,11 @@ phonenumbers =
 pydantic =
     pydantic>=1.8.1
 test =
-    mypy
+    mypy>=0.930
     pytest
-    pytest-mypy-plugins
+    # todo: Restore to mainline pytest-mypy-plugins once PR is closed.
+    #       https://github.com/typeddjango/pytest-mypy-plugins/pull/92
+    pytest-mypy-plugins @ https://github.com/antonagestam/pytest-mypy-plugins/archive/01957984e26118c9b176e3d055c858e1db029cf1.tar.gz
     coverage
 
 [flake8]

--- a/setup.cfg
+++ b/setup.cfg
@@ -39,9 +39,9 @@ phantom = py.typed
 
 [options.extras_require]
 phonenumbers =
-    phonenumbers>=8.12.11
+    phonenumbers>=8.12.41
 pydantic =
-    pydantic>=1.8.1
+    pydantic>=1.9.0
 test =
     mypy>=0.930
     pytest
@@ -94,11 +94,8 @@ disallow_any_expr = False
 disallow_untyped_decorators = False
 disallow_incomplete_defs = False
 
-[mypy-typeguard.*,phonenumbers.*]
+[mypy-typeguard.*]
 ignore_errors = True
-
-[mypy-setuptools,phonenumbers]
-ignore_missing_imports = True
 
 [coverage:run]
 source = src

--- a/src/phantom/base.py
+++ b/src/phantom/base.py
@@ -141,7 +141,7 @@ class Phantom(PhantomBase, Generic[T]):
       allows deferring definitions of ``predicate`` and ``bound`` to concrete subtypes.
     """
 
-    __predicate__: ClassVar[Predicate[object]]
+    __predicate__: Predicate[T]
     # The bound of a phantom type is the type that its values will have at
     # runtime, so when checking if a value is an instance of a phantom type,
     # it's first checked to be within its bounds, so that the value can be

--- a/src/phantom/base.py
+++ b/src/phantom/base.py
@@ -50,7 +50,7 @@ class PhantomMeta(abc.ABCMeta):
     # With the current level of metaclass support in mypy it's unlikely that we'll be
     # able to make this context typed, hence the ignores.
     def __call__(cls, instance):  # type: ignore[no-untyped-def]
-        return cls.parse(instance)  # type: ignore[attr-defined,misc]
+        return cls.parse(instance)  # type: ignore[attr-defined]
 
 
 class BoundError(TypeError):
@@ -141,7 +141,7 @@ class Phantom(PhantomBase, Generic[T]):
       allows deferring definitions of ``predicate`` and ``bound`` to concrete subtypes.
     """
 
-    __predicate__: ClassVar[Predicate[T]]
+    __predicate__: ClassVar[Predicate[object]]
     # The bound of a phantom type is the type that its values will have at
     # runtime, so when checking if a value is an instance of a phantom type,
     # it's first checked to be within its bounds, so that the value can be
@@ -159,7 +159,7 @@ class Phantom(PhantomBase, Generic[T]):
         abstract: bool = False,
         **kwargs: Any,
     ) -> None:
-        super().__init_subclass__(**kwargs)  # type: ignore[call-arg]
+        super().__init_subclass__(**kwargs)
         resolve_class_attr(cls, "__abstract__", abstract)
         resolve_class_attr(cls, "__predicate__", predicate)
         cls._resolve_bound(bound)

--- a/src/phantom/fn.py
+++ b/src/phantom/fn.py
@@ -33,10 +33,10 @@ def compose2(a: Callable[[AA], AR], b: Callable[[BA], AA]) -> Callable[[BA], AR]
     b_name = _name(b)
 
     def c(arg: BA) -> AR:
-        f"""Function composed as {a_name}({b_name}(_))."""
         return a(b(arg))
 
     c.__name__ = f"{a_name}∘{b_name}"
+    c.__doc__ = f"Function composed as {a_name}({b_name}(_))."
     return c
 
 

--- a/src/phantom/fn.py
+++ b/src/phantom/fn.py
@@ -33,7 +33,7 @@ def compose2(a: Callable[[AA], AR], b: Callable[[BA], AA]) -> Callable[[BA], AR]
     b_name = _name(b)
 
     def c(arg: BA) -> AR:
-        f"Function composed as {a_name}({b_name}(_))."
+        f"""Function composed as {a_name}({b_name}(_))."""
         return a(b(arg))
 
     c.__name__ = f"{a_name}∘{b_name}"

--- a/src/phantom/utils.py
+++ b/src/phantom/utils.py
@@ -136,6 +136,7 @@ def is_not_mutable(type_: BoundType) -> TypeGuard[NotKnownMutable]:
     return True
 
 
+MaybeUnionType: type | None
 try:
     MaybeUnionType = types.UnionType  # type: ignore[attr-defined]
 except AttributeError:
@@ -143,7 +144,7 @@ except AttributeError:
 
 
 def is_union_type(value: object) -> TypeGuard[type]:
-    return MaybeUnionType and isinstance(value, MaybeUnionType)
+    return MaybeUnionType is not None and isinstance(value, MaybeUnionType)
 
 
 def is_union(value: object) -> TypeGuard[type]:

--- a/tests/pydantic/test_pydantic_schemas.py
+++ b/tests/pydantic/test_pydantic_schemas.py
@@ -158,6 +158,7 @@ class TestShippedTypesImplementsSchema:
 
     def test_sized_non_empty_implements_schema(self):
         assert DataModel.schema()["properties"]["non_empty"] == {
+            "allOf": [{"type": "string"}],
             "title": "NonEmpty",
             "type": "array",
             "description": "A non-empty array.",

--- a/tests/test_fn.yaml
+++ b/tests/test_fn.yaml
@@ -1,24 +1,33 @@
 - case: test_compose2_can_infer_types
+  disable_cache: true
+  regex: true
   main: |
     from typing import Callable
     from phantom import Predicate
     from phantom.fn import compose2
     from predicates import is_valid_name
 
-    reveal_type(is_valid_name)  # N: Revealed type is "def (builtins.str) -> builtins.bool"
-    is_valid_name(1)  # E: Argument 1 has incompatible type "int"; expected "str"  [arg-type]
-    is_valid_name([])  # E: Argument 1 has incompatible type "List[<nothing>]"; expected "str"  [arg-type]
-    reveal_type(is_valid_name(""))  # N: Revealed type is "builtins.bool"
+    reveal_type(is_valid_name)
+    is_valid_name(1)
+    is_valid_name([])
+    reveal_type(is_valid_name(""))
     def takes_str_predicate(predicate: Predicate[str]) -> None: ...
     takes_str_predicate(is_valid_name)
 
-    reveal_type(compose2(int, str)(2))  # N: Revealed type is "builtins.int*"
+    reveal_type(compose2(int, str)(2))
     def takes_int_str_composed(fn: Callable[[int], int]) -> None: ...
     takes_int_str_composed(compose2(int, str))
 
-    reveal_type(compose2(str, int)("3"))  # N: Revealed type is "builtins.str*"
+    reveal_type(compose2(str, int)("3"))
     def takes_str_int_composed(fn: Callable[[str], str]) -> None: ...
     takes_str_int_composed(compose2(str, int))
+  out: |
+    main:6: note: Revealed type is "def \(builtins\.str\*?\) -> builtins\.bool\*?"
+    main:7: error: Argument 1 has incompatible type "int"; expected "str"  \[arg-type\]
+    main:8: error: Argument 1 has incompatible type "List\[<nothing>\]"; expected "str"  \[arg-type\]
+    main:9: note: Revealed type is "builtins.bool\*?"
+    main:13: note: Revealed type is "builtins.int\*?"
+    main:17: note: Revealed type is "builtins.str\*?"
   files:
     - path: predicates.py
       content: |

--- a/tests/test_fn.yaml
+++ b/tests/test_fn.yaml
@@ -5,10 +5,10 @@
     from phantom.fn import compose2
     from predicates import is_valid_name
 
-    reveal_type(is_valid_name)  # N: Revealed type is "def (builtins.str*) -> builtins.bool*"
+    reveal_type(is_valid_name)  # N: Revealed type is "def (builtins.str) -> builtins.bool"
     is_valid_name(1)  # E: Argument 1 has incompatible type "int"; expected "str"  [arg-type]
     is_valid_name([])  # E: Argument 1 has incompatible type "List[<nothing>]"; expected "str"  [arg-type]
-    reveal_type(is_valid_name(""))  # N: Revealed type is "builtins.bool*"
+    reveal_type(is_valid_name(""))  # N: Revealed type is "builtins.bool"
     def takes_str_predicate(predicate: Predicate[str]) -> None: ...
     takes_str_predicate(is_valid_name)
 

